File: packages/terminal-recovery/src/protocol.ts
export interface RecoverySnapshot {
  sessionId: string;
  serialized: string;
  cols: number;
  rows: number;
  savedAt: number;
  sequence: number;
}

export interface StartSessionCommand {
  type: "StartSession";
  sessionId: string;
  cols: number;
  rows: number;
}

export interface WriteOutputCommand {
  type: "WriteOutput";
  sessionId: string;
  data: number[];
  sequence: number;
}

export interface ResizeSessionCommand {
  type: "ResizeSession";
  sessionId: string;
  cols: number;
  rows: number;
}

export interface EndSessionCommand {
  type: "EndSession";
  sessionId: string;
}

export interface GetSnapshotCommand {
  type: "GetSnapshot";
  sessionId: string;
}

export interface FlushAndShutdownCommand {
  type: "FlushAndShutdown";
}

export type RecoveryCommand =
  | StartSessionCommand
  | WriteOutputCommand
  | ResizeSessionCommand
  | EndSessionCommand
  | GetSnapshotCommand
  | FlushAndShutdownCommand;

export interface OkResponse {
  type: "Ok";
}

export interface ErrorResponse {
  type: "Error";
  message: string;
}

export interface SnapshotResponse extends RecoverySnapshot {
  type: "Snapshot";
}

export interface NotFoundResponse {
  type: "NotFound";
}

export type RecoveryResponse =
  | OkResponse
  | ErrorResponse
  | SnapshotResponse
  | NotFoundResponse;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isString(value: unknown): value is string {
  return typeof value === "string";
}

function isNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isNumberArray(value: unknown): value is number[] {
  return Array.isArray(value) && value.every(isNumber);
}

export function parseCommand(line: string): RecoveryCommand {
  let parsed: unknown;
  try {
    parsed = JSON.parse(line);
  } catch (error) {
    throw new Error(`Invalid JSON: ${error instanceof Error ? error.message : String(error)}`);
  }

  if (!isRecord(parsed) || !isString(parsed.type)) {
    throw new Error("Invalid command: missing type");
  }

  switch (parsed.type) {
    case "StartSession":
      if (isString(parsed.sessionId) && isNumber(parsed.cols) && isNumber(parsed.rows)) {
        return {
          type: "StartSession",
          sessionId: parsed.sessionId,
          cols: parsed.cols,
          rows: parsed.rows,
        };
      }
      break;
    case "WriteOutput":
      if (isString(parsed.sessionId) && isNumberArray(parsed.data) && isNumber(parsed.sequence)) {
        return {
          type: "WriteOutput",
          sessionId: parsed.sessionId,
          data: parsed.data,
          sequence: parsed.sequence,
        };
      }
      break;
    case "ResizeSession":
      if (isString(parsed.sessionId) && isNumber(parsed.cols) && isNumber(parsed.rows)) {
        return {
          type: "ResizeSession",
          sessionId: parsed.sessionId,
          cols: parsed.cols,
          rows: parsed.rows,
        };
      }
      break;
    case "EndSession":
      if (isString(parsed.sessionId)) {
        return {
          type: "EndSession",
          sessionId: parsed.sessionId,
        };
      }
      break;
    case "GetSnapshot":
      if (isString(parsed.sessionId)) {
        return {
          type: "GetSnapshot",
          sessionId: parsed.sessionId,
        };
      }
      break;
    case "FlushAndShutdown":
      return { type: "FlushAndShutdown" };
    default:
      throw new Error(`Unknown command type: ${parsed.type}`);
  }

  throw new Error(`Invalid command payload for ${parsed.type}`);
}

export function formatResponse(response: RecoveryResponse): string {
  return `${JSON.stringify(response)}\n`;
}

File: packages/terminal-recovery/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RecoverySnapshot {
    pub session_id: String,
    pub serialized: String,
    pub cols: u16,
    pub rows: u16,
    pub cursor_row: u16,
    pub cursor_col: u16,
    pub cursor_visible: bool,
    pub saved_at: u64,
    pub sequence: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "type")]
pub enum RecoveryCommand {
    StartSession {
        #[serde(rename = "sessionId")]
        session_id: String,
        cols: u16,
        rows: u16,
        #[serde(rename = "resumeFromDisk")]
        #[serde(default)]
        resume_from_disk: bool,
    },
    WriteOutput {
        #[serde(rename = "sessionId")]
        session_id: String,
        data: Vec<u8>,
        sequence: u64,
    },
    ResizeSession {
        #[serde(rename = "sessionId")]
        session_id: String,
        cols: u16,
        rows: u16,
    },
    EndSession {
        #[serde(rename = "sessionId")]
        session_id: String,
    },
    GetSnapshot {
        #[serde(rename = "sessionId")]
        session_id: String,
    },
    FlushAndShutdown,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "type")]
pub enum RecoveryResponse {
    Ok,
    Error {
        message: String,
    },
    Snapshot {
        #[serde(rename = "sessionId")]
        session_id: String,
        serialized: String,
        cols: u16,
        rows: u16,
        #[serde(rename = "cursorRow")]
        cursor_row: u16,
        #[serde(rename = "cursorCol")]
        cursor_col: u16,
        #[serde(rename = "cursorVisible")]
        cursor_visible: bool,
        #[serde(rename = "savedAt")]
        saved_at: u64,
        sequence: u64,
    },
    NotFound,
}

impl RecoveryResponse {
    pub fn from_snapshot(snapshot: RecoverySnapshot) -> Self {
        Self::Snapshot {
            session_id: snapshot.session_id,
            serialized: snapshot.serialized,
            cols: snapshot.cols,
            rows: snapshot.rows,
            cursor_row: snapshot.cursor_row,
            cursor_col: snapshot.cursor_col,
            cursor_visible: snapshot.cursor_visible,
            saved_at: snapshot.saved_at,
            sequence: snapshot.sequence,
        }
    }
}

pub fn parse_command(line: &str) -> Result<RecoveryCommand, String> {
    serde_json::from_str(line).map_err(|error| format!("Invalid JSON: {}", error))
}

pub fn format_response(response: &RecoveryResponse) -> Result<String, String> {
    let mut json = serde_json::to_string(response)
        .map_err(|error| format!("failed to serialize response: {}", error))?;
    json.push('\n');
    Ok(json)
}
